src/config/mod.rs

```diff
@@ -117,13 +118,61 @@ pub struct MiddlewaresConfig {
     pub subscriptions: Vec<String>,
 }
 
-#[derive(Debug)]
+#[derive(Debug, Validate)]
+#[garde(allow_unvalidated)]
 pub struct Config {
+    #[garde(dive)]
     pub extensions: ExtensionsConfig,
+    #[garde(custom(has_matched_extensions(&self.extensions)))]
     pub middlewares: MiddlewaresConfig,
+    #[garde(dive)]
     pub rpcs: RpcDefinitions,
 }
 
+macro_rules! define_middleware_extension_mappings {
+    (
+        methods: [$(($method_mw:ident, $method_ext:ident)),*],
+        subscriptions: [$(($sub_mw:ident, $sub_ext:ident)),*]
+    ) => {
+        fn has_matched_extensions(extensions: &ExtensionsConfig) -> impl FnOnce(&MiddlewaresConfig, &()) -> garde::Result + '_ {
+            move |middlewares, _| {
+                for m in &middlewares.methods {
+                    match m.as_str() {
+                        $(
+                            stringify!($method_mw) => {
+                                if extensions.$method_ext.is_none() {
+                                    return Err(garde::Error::new(format!("No extension: {} for methods middleware: {} in the config. Please check the config.", stringify!($method_ext), stringify!($method_mw))));
+                                }
+                            }
+                        )*
+                        // We don't care the middleware that doesn't have matching extension
+                        _ => {}
+                    }
+                }
+                for s in &middlewares.subscriptions {
+                    match s.as_str() {
+                        $(
+                            stringify!($sub_mw) => {
+                                if extensions.$sub_ext.is_none() {
+                                    return Err(garde::Error::new(format!("No extension: {} for subscriptions middleware: {} in the config. Please check the config.", stringify!($sub_ext), stringify!($sub_mw))));
+                                }
+                            }
+                        )*
+                        // We don't care the middleware that doesn't have matching extension
+                        _ => {}
+                    }
+                }
+                Ok(())
+            }
+        }
+    };
+}
+
+define_middleware_extension_mappings! {
```

Comment:
this isn't very extensible as we need to define all the mapping here and it is very easy to make mistake when we introduce new extensions or make modifications.

should add a new `validate` method to the extension and call it instead. in that way the validation and use are in a same file and harder to modify one without modify others. and we should be able to do it in such way to ensure it is not possible to require an extension without validate it.

Reply:
Yeah, I agree. As you said, to make it not possible to use without validating it, I think the best way is to ensure it via type-safety. The process is like this: first we validate at some entrypoint which typically in the very beginning, then we wrap it into a new-type, and use this validated new-type all over the rest. At everywhere we use the new-type, we know the value has been validated. But it may take lots of changes in the codes, I will try to get a workaround.

Reply:
let's remove this part for now so it is not blocking this PR and then we can do the extension validation in next PR

Reply:
All right.